### ISSUE TITLE
 Zlib: deflateEnd() fails after deflateInit2()/deflateReset() 

### ISSUE THREAD

**Issue by 1234sv** [zlib-ng/zlib-ng#123]
This duplicate of issue #250 created under madler project here https://github.com/madler/zlib/issues/250.

In summary, following call sequence fails for "Raw deflate i.e. deflateInit() done with window bits -15":
deflateInit() -->deflateEnd().
deflateInit() --> deflateReset() -->deflateEnd().

More details on link above.

**Comment by mtl1979** [comment]
Invalid example. Can be closed.

**Comment by 1234sv** [comment]
how come this is an invalid example. Can you prove me?
A simple sequence of calling deflateInit() --> deflateEnd() should work. Secondly,
deflateInit --> deflateReset --> deflateEnd() should also work.
In issue https://github.com/madler/zlib/issues/250, i described even 3rd example. 
If you could prove these are invalid example, that would make better sense to me,

**Comment by mtl1979** [comment]
@1234sv The second example is missing last Z_FINISH outside the loop. Like I already mentioned there, Z_FINISH is only supposed to be used right before deflateEnd(). deflate() already knows when the stream has ended so there is no reason to use Z_FINISH inside the loop. When deflate() returns no more bytes, the stream has ended, so one only needs to check how many bytes there is in output buffer.

**Comment by 1234sv** [comment]
to simplyfy it .. its like:

deflateInit()
deflate(Z_FINISH)
deflateReset();
deflateEnd();

So, is this a valid call flow?

**Comment by mtl1979** [comment]
@1234sv Essentially the flow should be (simplified):

deflateInit();
deflate(Z_FINISH);
deflateReset();
deflate(Z_FINISH);
deflateReset();
deflate(Z_FINISH);
deflateReset();
deflate(Z_FINISH);
deflateReset();
deflate(Z_FINISH);
deflateReset();
deflate(Z_FINISH);
deflateReset();
deflate(Z_FINISH);
deflateReset();
deflate(Z_FINISH);
deflateReset();
deflate(Z_FINISH);
deflateReset();
deflate(Z_FINISH);
deflateEnd();

So, you need to skip deflateReset() call on last iteration of the loop.

**Comment by 1234sv** [comment]
why that's the dependency?

**Comment by mtl1979** [comment]
@1234sv Most likely to warn user about truncated output. When deflate is used with Z_FINISH and big enough buffer, the output is guaranteed to be full length. deflateReset() will - like the name suggests - reset the result of sanity checks. When deflate is used with Z_FINISH but too small buffer, it will return Z_OK or Z_BUF_ERROR instead of Z_STREAM_END.

**Comment by 1234sv** [comment]
"When deflate is used with Z_FINISH but too small buffer, " -- in this case, deflateEnd() too will fail as zlib is in middle of operation. so a simple sequence like this: 
deflateinit()
deflate(Z_FINISH); // buf is just 1 byte
deflateEnd(); // --> this will fail regardless.

so, this explanation doesn't explain situation why a Reset cannot be called before End().

Anyways, if I further simply my question that its 
why follow call flow cannot work

deflateInit() 
deflateEnd(). ?

**Comment by mtl1979** [comment]
deflateReset() cannot be called before deflateEnd() because deflateEnd() depends on deflate state that deflateReset() doesn't preserve.
Simple deflateInit() -> deflateEnd() doesn't work because deflateInit() doesn't actually output anything, first deflate() call does the actual work.

**Comment by 1234sv** [comment]
Simple deflateInit() -> deflateEnd() doesn't work because deflateInit() doesn't actually output anything, first deflate() call does the actual work.

>> but it works if we open stream for ZLIB decode. and it should work same of raw deflate as well

**Comment by mtl1979** [comment]
@1234sv deflateInit() and deflateInit2() internally call deflateReset() so that is why it doesn't work for raw deflate. The state is undefined until first deflate() call. It doesn't matter how many times user or zlib code calls deflateReset() in raw mode, the state is always undefined.

**Comment by 1234sv** [comment]
So, what is ideal sequence of operation for Raw deflate?

**Comment by mtl1979** [comment]
For ideal sequence, see https://github.com/Dead2/zlib-ng/issues/123#issuecomment-328253511

**Comment by 1234sv** [comment]
I understand your input how zlib implementation for Raw deflate works and that's what my observation is as well. However,
 I still dont understand why user is forced to call deflate(Z_FINISH) *just* before End() as user can always choose to End operation anytime , say ctrl+c pressed or any other situation. And why user *just* cannot call deflateEnd() if it is already reset. As a user, I see algorithm should be allow user to close it anytime by not putting some hard requirements on sequencing. So, following some specific call flow may result in some erroneous results but API should allow it.

**Comment by mtl1979** [comment]
@1234sv I think Mark Adler can explain the reason behind the logic better than I... I only know what I have read from the manual or the source tree.

**Comment by 1234sv** [comment]
Waiting on Mark to look at this. Thanks for your input on this issue.

**Comment by Dead2** [comment]
Looks like this was fixed upstream.
